_posts/2022-10-18-torch-linalg/torch-linalg.Rmd

```diff
@@ -0,0 +1,632 @@
+
+$$
+\mathbf{A}^T(\mathbf{Ax} - \mathbf{b}) = \mathbf{0}
+$$
+
+A rearrangement of this equation yields the so-called *normal equations*:
+
+$$
+\mathbf{A}^T \mathbf{A} \mathbf{x} = \mathbf{A}^T \mathbf{b}
+$$
+
+These may be solved for $\mathbf{x}$, computing the inverse of $\mathbf{A}^T\mathbf{A}$:
+
+$$
+\mathbf{x} = (\mathbf{A}^T \mathbf{A})^{-1} \mathbf{A}^T \mathbf{b}
+$$
+
+$\mathbf{A}^T\mathbf{A}$ is a square matrix. It still might not be invertible, in which case the so-called pseudoinverse would be computed instead. In our case, this will not be needed; we already know $\mathbf{A}$ has full rank, and so does $\mathbf{A}^T\mathbf{A}$.
+
+Thus, from the normal equations we have derived a recipe for computing $\mathbf{b}$. Let's put it to use, and compare with what we got from `lm()` and `linalg_lstsq()`.
+
+```{r}
+AtA <- A$t()$matmul(A)
+Atb <- A$t()$matmul(b)
+inv <- linalg_inv(AtA)
+x <- inv$matmul(Atb)
+
+all_preds$neq <- as.matrix(A$matmul(x))
+all_errs$neq <- rmse(all_preds$b, all_preds$neq)
+
+all_errs
+
+```
+
+           lm   lstsq     neq
+    1 40.8369 40.8369 40.8369
+
+Having confirmed that the direct way works, we may allow ourselves some sophistication. Four different matrix factorizations will make their appearance: Cholesky, LU, QR, and Singular Value Decomposition. The goal, in every case, is to avoid the expensive computation of the (pseudo-) inverse. That's what all methods have in common. However, they do not differ "just" in the way the matrix is factorized, but also, in *which* matrix is. This has to do with the constraints the various methods impose. Roughly speaking, the order they're listed in above reflects a falling slope of preconditions, or put differently, a rising slope of generality. Due to the constraints involved, the first two (Cholesky, as well as LU decomposition) will be performed on $\mathbf{A}^T\mathbf{A}$, while the latter two (QR and SVD) operate on $\mathbf{A}$ directly. With them, there never is a need to compute $\mathbf{A}^T\mathbf{A}$.
+
+## Least squares (II): Cholesky decomposition
+
+In Cholesky decomposition, a matrix is factored into two triangular matrices of the same size, with one being the transpose of the other. This commonly is written either
+
+$$
+\mathbf{A} = \mathbf{L} \mathbf{L}^T
+$$ or
+
+$$
+\mathbf{A} = \mathbf{R}^T\mathbf{R}
+$$
+
+Here symbols $\mathbf{L}$ and $\mathbf{R}$ denote lower-triangular and upper-triangular matrices, respectively.
+
+For Cholesky decomposition to be possible, a matrix has to be both symmetric and positive definite. These are pretty strong conditions, ones that will not often be fulfilled in practice. In our case, $\mathbf{A}$ is not symmetric; this immediately implies we have to operate on $\mathbf{A}^T\mathbf{A}$ instead. And since $\mathbf{A}$ already is positive definite, we know that $\mathbf{A}^T\mathbf{A}$ is, as well.
```

Comment:
```suggestion
For Cholesky decomposition to be possible, a matrix has to be both symmetric and positive definite. These are pretty strong conditions, ones that will not often be fulfilled in practice. In our case, $\mathbf{A}$ is not symmetric. This immediately implies we have to operate on $\mathbf{A}^T\mathbf{A}$ instead. And since $\mathbf{A}$ already is positive definite, we know that $\mathbf{A}^T\mathbf{A}$ is, as well.
```